timesketch/ui/graphs/cytoscape.component.ts

```diff
@@ -0,0 +1,224 @@
+  @Output() vmouseup = new EventEmitter<Evt['mouseup' | 'touchend']>()
+  @Output() tap = new EventEmitter<Evt['click' | 'touchend']>()
+  @Output() vclick = new EventEmitter<Evt['click' | 'touchend']>()
+  @Output() taphold = new EventEmitter<Evt['taphold']>()
+  @Output() cxttapstart = new EventEmitter<Evt['mousedown' | 'touchstart']>()
+  @Output() cxttapend = new EventEmitter<Evt['mouseup' | 'touchend']>()
+  @Output() cxttap = new EventEmitter<Evt['click' | 'touchend']>()
+  @Output() cxtdrag = new EventEmitter<Evt['touchmove' | 'mousemove']>()
+  @Output() cxtdragover = new EventEmitter<Evt['touchmove' | 'mousemove' | 'mouseover']>()
+  @Output() cxtdragout = new EventEmitter<Evt['touchmove' | 'mousemove' | 'mouseout']>()
+  @Output() boxstart = new EventEmitter<Evt['boxstart']>()
+  @Output() boxend = new EventEmitter<Evt['boxend']>()
+  @Output() boxselect = new EventEmitter<Evt['boxselect']>()
+  @Output() box = new EventEmitter<Evt['box']>()
+
+  // graph events
+  @Output() layoutstart = new EventEmitter<Evt['layoutstart']>()
+  @Output() layoutready = new EventEmitter<Evt['layoutready']>()
+  @Output() layoutstop = new EventEmitter<Evt['layoutstop']>()
+  @Output() ready = new EventEmitter<Evt['ready']>()
+  @Output() destroy = new EventEmitter<Evt['destroy']>()
+  @Output() render = new EventEmitter<Evt['render']>()
+  @Output() resize = new EventEmitter<Evt['resize']>()
+
+  // tslint:enable:no-unused-variable
+
+  constructor(el: ElementRef) {
+    this.element = el.nativeElement
+  }
+
+  initCytoscape() {
+    const all_options = [].concat(immutable_options, mutable_options)
+    const options: Opt = {}
+    if (this.element) {
+      options['container'] = this.element.childNodes[0]
+    }
+    for (const k of all_options) {
+      options[k] = this[k]
+    }
+    this._cy = cytoscape(options)
+    this.initEvents()
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    const changed = (k) =>
+      changes[k].previousValue !== changes[k].currentValue
+    const options: Opt = {}
+    let needs_rebuild = false
+    for (const k of Object.keys(changes)) {
+      if (changed(k)) {
+        options[k] = changes[k].currentValue
+      }
+      if (immutable_options.indexOf(k) > -1 && this.cy) {
+        needs_rebuild = true
+        break
+      }
+    }
+    if (this.cy) {
+      if (needs_rebuild) {
+        this.cy.destroy()
+        this._cy = undefined
+        this.initCytoscape()
+      } else {
+        (this.cy.json as any)(options)
+      }
+    }
+  }
+
+  ngAfterViewInit() {
+    this.initCytoscape()
+  }
+
+  private initEvents() {
+    this.cy.on('pan', () => {
+      this.pan = this.cy.pan()
+      this.panChange.emit(this.pan)
+    })
+    this.cy.on('zoom', () => {
+      this.zoom = this.cy.zoom()
+      this.zoomChange.emit(this.zoom)
+    })
+    for (const [k, v] of Object.entries(this)) {
+      if (v instanceof EventEmitter && !k.endsWith('Change')) {
```

Comment:
This is again a bit fragile. You do not document that an EventEmitter without Change has a special meaning. I would either switch to a map from string to EventEmitter or make it very clear how to add new fields.

Reply:
I updated the documentation. It is intended that this component mirrors Cytoscape API as closely as possible, without any custom additions.